### PR TITLE
Setting reliable archive mode EN

### PR DESCRIPTION
**Pull request by sjnam**
## Proposed changes

- Full EN can only store the latest block states (e.g. latest-128 block ~ latest block)
- Querying latest block state of archive EN is not efficient compared to query latest block state of full EN
- You can add the command line flag of an EN `--upstream-en archive-en-rpc-url` to test. For example, `--upstream-en http://localhost:8557`

## Types of changes

- [ ] Bugfix
- [x] New feature or enhancement
- [ ] Others

## Checklist

- [x] I have read the [CONTRIBUTING GUIDELINES](https://github.com/klaytn/klaytn/blob/master/CONTRIBUTING.md) doc
- [x] I have signed the [CLA](https://cla-assistant.io/klaytn/klaytn)
- [x] Lint and unit tests pass locally with my changes (`$ make test`)
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged and published in downstream modules

## Further comments

If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: networks/rpc/handler.go [hunk 1]
```diff
@@ -427,6 +428,23 @@ func (h *handler) handleSubscribe(cp *callProc, msg *jsonrpcMessage) *jsonrpcMes
 func (h *handler) runMethod(ctx context.Context, msg *jsonrpcMessage, callb *callback, args []reflect.Value) *jsonrpcMessage {
 	result, err := callb.call(ctx, msg.Method, args)
 	if err != nil {
+		if _, ok := err.(*statedb.MissingNodeError); ok {
```
Comment (blukat29) [inline networks/rpc/handler.go:431]:
```go
if UpstreamArchiveEN != "" && shouldRequestUpstream(result, err) {
    return requestUpstream(ctx, msg, args)
}
```

```go
func shouldRequestUpstream(result, err) bool {
    switch err.(type) {
    case *statedb.MissingNodeError:
        return true
    default:
        return false
    }
}

func requestUpstream(ctx, msg, args) *jsonrpcMessage {
}
```

Comment (sjnam) [reply networks/rpc/handler.go:431]: @blukat29 
`UpstreamArchiveEN != ""` is checked at `url.Parse(rawurl)` of `DialContext` method.
https://github.com/sjnam/klaytn/blob/reliable-archive-en/networks/rpc/client.go#L176-L180
But, it's good idea to explicitly check for `UpstreamArchiveEN != ""`.

Comment (blukat29) [reply networks/rpc/handler.go]: @sjnam I see. But if we dial when `UpstreamArchiveEN == ""`, then (depending on the implementation) the outer RPC returns "URL parse error", not "missing node error". So I think the explicit check is necessary.

**Review by blukat29 — APPROVED** [review]

File: networks/rpc/handler.go [hunk 2]
```diff
@@ -427,6 +428,9 @@ func (h *handler) handleSubscribe(cp *callProc, msg *jsonrpcMessage) *jsonrpcMes
 func (h *handler) runMethod(ctx context.Context, msg *jsonrpcMessage, callb *callback, args []reflect.Value) *jsonrpcMessage {
 	result, err := callb.call(ctx, msg.Method, args)
 	if err != nil {
+		if UpstreamArchiveEN != "" && shouldRequestUpstream(err) {
```
Comment (hyunsooda) [inline networks/rpc/handler.go:431]: LGTM. Regarding this feature, let me ask three usage questions.

1. The single URL may be extended to the list of URL. Was there an issue regarding this?
2. The URL (list or single) seems not modifiable at runtime. If the initially set node becomes unreachable, the routing node should be restarted to change that value. It seems to degrade the usage.
3. Any request can be relayed as well as the state lack. Is this an obvious limitation by the consideration of big load to the upstream node?

Comment (hyunsooda) [reply networks/rpc/handler.go:431]: Got it thanks.

**Review by hyunsooda — DISMISSED** [review]

File: networks/rpc/handler.go (hunk 2 above)
Comment (yoomee1313) [reply networks/rpc/handler.go:431]: @sjnam can you add todo about those suggestions? Also, not mentioned here, additional suggestion is to make a new rpcErrorResponse struct for it.

**Review by blukat29 — APPROVED** [review]

**Comment by blukat29** [comment]
@sjnam Please rebase against latest dev.

**Review by yoomee1313 — APPROVED** [review]

**Pull request merged**